File: go/fory/serializer.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package fory

import (
	"reflect"
)

// Serializer is the unified interface for all serialization.
// It provides reflect.Value-based API for efficient serialization.
// All methods use centralized error handling via context - errors are set on
// the context and checked at strategic points (entry/exit of complex types).
type Serializer interface {
	// Write is the entry point for serialization.
	//
	// This method orchestrates the complete serialization process, handling reference tracking,
	// type information, and delegating to WriteData for the actual data serialization.
	//
	// Unlike Java's unified ref tracking approach, Go uses per-serializer ref/type handling.
	// Map, Slice, Interface, and Pointer types each have different ref tracking requirements,
	// so each serializer controls how to write ref/type info. This allows more efficient
	// serialization by avoiding unnecessary ref checks for value types and enabling
	// type-specific optimizations.
	//
	// Parameters:
	//   - refMode: controls reference/null handling behavior:
	//     - RefModeNone: skip ref handling entirely
	//     - RefModeNullOnly: only write null flag (NullFlag or NotNullValueFlag)
	//     - RefModeTracking: full reference tracking with WriteRefOrNull
	//   - writeType: when true, writes type information; when false, skips it
	//   - hasGenerics: when true, indicates element types are known from TypeDef (struct field context),
	//     so container serializers can skip writing element type info
	//
	// Errors are set on the context via ctx.SetError() and should be checked
	// at appropriate boundaries using ctx.HasError() or ctx.CheckError().
	Write(ctx *WriteContext, refMode RefMode, writeType bool, hasGenerics bool, value reflect.Value)

	// WriteData serializes using reflect.Value.
	// Does NOT write ref/type info - caller handles that.
	// Errors are set on the context via ctx.SetError().
	WriteData(ctx *WriteContext, value reflect.Value)

	// Read is the entry point for deserialization.
	//
	// This method orchestrates the complete deserialization process, handling reference tracking,
	// type information validation, and delegating to ReadData for the actual data deserialization.
	//
	// Unlike Java's unified ref tracking approach, Go uses per-serializer ref/type handling.
	// Map, Slice, Interface, and Pointer types each have different ref tracking requirements,
	// so each serializer controls how to read ref/type info. This allows more efficient
	// deserialization by avoiding unnecessary ref checks for value types and enabling
	// type-specific optimizations.
	//
	// Parameters:
	//   - refMode: controls reference/null handling behavior:
	//     - RefModeNone: skip ref handling entirely
	//     - RefModeNullOnly: only read null flag
	//     - RefModeTracking: full reference tracking with TryPreserveRefId
	//   - readType: when true, reads type information from buffer; when false, skips it
	//   - hasGenerics: when true, indicates element types are known from TypeDef (struct field context),
	//     so container serializers can skip reading element type info
	//
	// Errors are set on the context via ctx.SetError() and should be checked
	// at appropriate boundaries using ctx.HasError() or ctx.CheckError().
	Read(ctx *ReadContext, refMode RefMode, readType bool, hasGenerics bool, value reflect.Value)

	// ReadData deserializes directly into the provided reflect.Value.
	// Does NOT read ref/type info - caller handles that.
	// For non-trivial types (slices, maps), implementations should reuse existing capacity when possible.
	// This method should ONLY be used by collection serializers for nested element deserialization.
	// For general deserialization, use ReadFull instead.
	// Errors are set on the context via ctx.SetError().
	ReadData(ctx *ReadContext, value reflect.Value)

	// ReadWithTypeInfo deserializes with pre-read type information.
	//
	// This method is used when type information has already been read from the buffer
	// and needs to be passed to the deserialization logic. This is common in polymorphic
	// deserialization scenarios where the runtime type differs from the static type.
	//
	// Parameters:
	//   - refMode: controls reference/null handling behavior:
	//     - RefModeNone: skip ref handling entirely
	//     - RefModeNullOnly: only read null flag
	//     - RefModeTracking: full reference tracking with TryPreserveRefId
	//   - typeInfo: pre-read type information; do NOT read type info again from buffer
	//
	// Important: do NOT read type info from the buffer in this method. The typeInfo
	// parameter contains the already-read type metadata. Reading it again will cause
	// buffer position errors.
	// Errors are set on the context via ctx.SetError().
	ReadWithTypeInfo(ctx *ReadContext, refMode RefMode, typeInfo *TypeInfo, value reflect.Value)
}

// ExtensionSerializer is a simplified interface for user-implemented extension serializers.
// Users implement this interface to provide custom serialization logic for types
// registered via RegisterNamedExtension.
//
// Unlike the full Serializer interface, ExtensionSerializer only requires implementing
// the core data serialization logic - reference tracking, type info, and protocol
// details are handled automatically by Fory.
//
// Example:
//
//	type MyExtSerializer struct{}
//
//	func (s *MyExtSerializer) WriteData(ctx *WriteContext, value reflect.Value) {
//	    myExt := value.Interface().(MyExt)
//	    ctx.Buffer().WriteVarint32(myExt.Id)
//	}
//
//	func (s *MyExtSerializer) ReadData(ctx *ReadContext, value reflect.Value) {
//	    id := ctx.Buffer().ReadVarint32(ctx.Err())
//	    value.Set(reflect.ValueOf(MyExt{Id: id}))
//	}
//
//	// Register with custom serializer
//	f.RegisterNamedExtension(MyExt{}, "my_ext", &MyExtSerializer{})
type ExtensionSerializer interface {
	// WriteData serializes the value's data to the buffer.
	// Only write the data fields - don't write ref flags or type info.
	// Errors should be set on the context via ctx.SetError().
	WriteData(ctx *WriteContext, value reflect.Value)

	// ReadData deserializes the value's data from the buffer into the provided value.
	// Only read the data fields - don't read ref flags or type info.
	// Errors should be set on the context via ctx.SetError().
	ReadData(ctx *ReadContext, value reflect.Value)
}

File: java/fory-core/src/main/java/org/apache/fory/serializer/Serializer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.fory.serializer;

import javax.annotation.concurrent.NotThreadSafe;
import org.apache.fory.Fory;
import org.apache.fory.memory.MemoryBuffer;
import org.apache.fory.resolver.RefMode;
import org.apache.fory.resolver.RefResolver;
import org.apache.fory.type.TypeUtils;

/**
 * Serialize/deserializer objects into binary. Note that this class is designed as an abstract class
 * instead of interface to reduce virtual method call cost of {@link #needToWriteRef}.
 *
 * @param <T> type of objects being serializing/deserializing
 */
@NotThreadSafe
@SuppressWarnings("unchecked")
public abstract class Serializer<T> {
  protected final Fory fory;
  protected final RefResolver refResolver;
  protected final Class<T> type;
  protected final boolean isJava;
  protected final boolean needToWriteRef;

  /**
   * Whether to enable circular reference of copy. Only for mutable objects, immutable objects just
   * return itself.
   */
  protected final boolean needToCopyRef;

  protected final boolean immutable;

  public Serializer(Fory fory, Class<T> type) {
    this(
        fory,
        type,
        fory.trackingRef() && !TypeUtils.isBoxed(TypeUtils.wrap(type)),
        TypeUtils.isPrimitive(type) || TypeUtils.isBoxed(type));
  }

  public Serializer(Fory fory, Class<T> type, boolean immutable) {
    this(fory, type, fory.trackingRef() && !TypeUtils.isBoxed(TypeUtils.wrap(type)), immutable);
  }

  public Serializer(Fory fory, Class<T> type, boolean needToWriteRef, boolean immutable) {
    this.fory = fory;
    this.refResolver = fory.getRefResolver();
    this.type = type;
    this.isJava = !fory.isCrossLanguage();
    this.needToWriteRef = needToWriteRef;
    this.needToCopyRef = fory.copyTrackingRef() && !immutable;
    this.immutable = immutable;
  }

  /**
   * Write value to buffer, this method may write ref/null flags based passed {@code refMode}, and
   * the passed value can be null. Note that this method don't write type info, this method is
   * mostly be used in cases the context has already knows the value type when deserialization.
   */
  public void write(MemoryBuffer buffer, RefMode refMode, T value) {
    // noinspection Duplicates
    if (refMode == RefMode.TRACKING) {
      if (refResolver.writeRefOrNull(buffer, value)) {
        return;
      }
    } else if (refMode == RefMode.NULL_ONLY) {
      if (value == null) {
        buffer.writeByte(Fory.NULL_FLAG);
        return;
      } else {
        buffer.writeByte(Fory.NOT_NULL_VALUE_FLAG);
      }
    }
    write(buffer, value);
  }

  /**
   * Write value to buffer, this method do not write ref/null flags and the passed value must not be
   * null.
   */
  public void write(MemoryBuffer buffer, T value) {
    throw new UnsupportedOperationException("Please implement serialization for " + type);
  }

  /**
   * Read value from buffer, this method may read ref/null flags based passed {@code refMode}, and
   * the read value can be null. Note that this method don't read type info, this method is mostly
   * be used in cases the context has already knows the value type for deserialization.
   */
  public T read(MemoryBuffer buffer, RefMode refMode) {
    if (refMode == RefMode.TRACKING) {
      T obj;
      int nextReadRefId = refResolver.tryPreserveRefId(buffer);
      if (nextReadRefId >= Fory.NOT_NULL_VALUE_FLAG) {
        obj = read(buffer);
        refResolver.setReadObject(nextReadRefId, obj);
        return obj;
      } else {
        return (T) refResolver.getReadObject();
      }
    } else if (refMode != RefMode.NULL_ONLY || buffer.readByte() != Fory.NULL_FLAG) {
      if (needToWriteRef) {
        // in normal case, the xread implementation may invoke `refResolver.reference` to
        // support circular reference, so we still need this `-1`
        refResolver.preserveRefId(-1);
      }
      return read(buffer);
    }
    return null;
  }

  /**
   * Read value from buffer, this method wont read ref/null flags and the read value won't be null.
   */
  public T read(MemoryBuffer buffer) {
    throw new UnsupportedOperationException("Please implement serialization for " + type);
  }

  /**
   * Write value to buffer, this method may write ref/null flags based passed {@code refMode}, and
   * the passed value can be null. Note that this method don't write type info, this method is
   * mostly be used in cases the context has already knows the value type when deserialization.
   */
  public void xwrite(MemoryBuffer buffer, RefMode refMode, T value) {
    // noinspection Duplicates
    if (refMode == RefMode.TRACKING) {
      if (refResolver.writeRefOrNull(buffer, value)) {
        return;
      }
    } else if (refMode == RefMode.NULL_ONLY) {
      if (value == null) {
        buffer.writeByte(Fory.NULL_FLAG);
        return;
      } else {
        buffer.writeByte(Fory.NOT_NULL_VALUE_FLAG);
      }
    }
    xwrite(buffer, value);
  }

  /**
   * Write value to buffer, this method do not write ref/null flags and the passed value must not be
   * null.
   */
  public void xwrite(MemoryBuffer buffer, T value) {
    throw new UnsupportedOperationException("Please implement xlang serialization for " + type);
  }

  /**
   * Read value from buffer, this method may read ref/null flags based passed {@code refMode}, and
   * the read value can be null. Note that this method don't read type info, this method is mostly
   * be used in cases the context has already knows the value type for deserialization.
   */
  public T xread(MemoryBuffer buffer, RefMode refMode) {
    if (refMode == RefMode.TRACKING) {
      T obj;
      int nextReadRefId = refResolver.tryPreserveRefId(buffer);
      if (nextReadRefId >= Fory.NOT_NULL_VALUE_FLAG) {
        obj = xread(buffer);
        refResolver.setReadObject(nextReadRefId, obj);
        return obj;
      } else {
        return (T) refResolver.getReadObject();
      }
    } else if (refMode != RefMode.NULL_ONLY || buffer.readByte() != Fory.NULL_FLAG) {
      if (needToWriteRef) {
        // in normal case, the xread implementation may invoke `refResolver.reference` to
        // support circular reference, so we still need this `-1`
        refResolver.preserveRefId(-1);
      }
      return xread(buffer);
    }
    return null;
  }

  /**
   * Read value from buffer, this method won't read ref/null flags and the read value won't be null.
   */
  public T xread(MemoryBuffer buffer) {
    throw new UnsupportedOperationException("Please implement xlang serialization for " + type);
  }

  public T copy(T value) {
    if (isImmutable()) {
      return value;
    }
    throw new UnsupportedOperationException(
        String.format("Copy for %s is not supported", value.getClass()));
  }

  public final boolean needToWriteRef() {
    return needToWriteRef;
  }

  public final boolean needToCopyRef() {
    return needToCopyRef;
  }

  public Class<T> getType() {
    return type;
  }

  public boolean isImmutable() {
    return immutable;
  }
}
